From: Christoph Reiter <reiter.christoph@gmail.com>
Subject: [PATCH] cygwin: default to "spawn" for multiprocessing and disable
 "forkserver"

passing fds doesn't seem to be supported by cygwin, so the "forkserver"
variant of multiprocessing fails.

Since 3.14 "forkserver" is the default where HAVE_SEND_HANDLE is true, so
marking it as unsupported also makes "spawn" the default. Unlike earlier
versions get_all_start_methods() lists every concrete context, so also
remove it from there.
---
diff --git a/Lib/multiprocessing/context.py b/Lib/multiprocessing/context.py
index 7226be7..56826a4 100644
--- a/Lib/multiprocessing/context.py
+++ b/Lib/multiprocessing/context.py
@@ -335,6 +335,8 @@ if sys.platform != 'win32':
         'spawn': SpawnContext(),
         'forkserver': ForkServerContext(),
     }
+    if not reduction.HAVE_SEND_HANDLE:
+        del _concrete_contexts['forkserver']
 
     def _get_platform_default_context():
         # bpo-33725: running arbitrary code after fork() is no longer reliable
diff --git a/Lib/multiprocessing/reduction.py b/Lib/multiprocessing/reduction.py
index fcccd3e..0f78ea9 100644
--- a/Lib/multiprocessing/reduction.py
+++ b/Lib/multiprocessing/reduction.py
@@ -24,7 +24,7 @@ __all__ = ['send_handle', 'recv_handle', 'ForkingPickler', 'register', 'dump']
 HAVE_SEND_HANDLE = (sys.platform == 'win32' or
                     (hasattr(socket, 'CMSG_LEN') and
                      hasattr(socket, 'SCM_RIGHTS') and
-                     hasattr(socket.socket, 'sendmsg')))
+                     hasattr(socket.socket, 'sendmsg'))) and sys.platform != 'cygwin'
 
 #
 # Pickler subclass
